cirq/sim/simulator.py

```diff
@@ -471,29 +463,43 @@ def sample_measurement_ops(self,
                 instances or a qubit is acted upon multiple times by different
                 operations from `measurement_ops`.
         """
-        bounds = {}  # type: Dict[str, Tuple]
-        all_qubits = []  # type: List[ops.Qid]
-        meas_ops = {}
-        current_index = 0
+
+        # Sanity checks.
+        seen_measurement_keys: Set[str] = set()
         for op in measurement_ops:
             gate = op.gate
             if not isinstance(gate, ops.MeasurementGate):
-                raise ValueError('{} was not a MeasurementGate'.format(gate))
+                raise ValueError(f'{op.gate} was not a MeasurementGate')
             key = protocols.measurement_key(gate)
-            meas_ops[key] = gate
-            if key in bounds:
-                raise ValueError(
-                    'Duplicate MeasurementGate with key {}'.format(key))
-            bounds[key] = (current_index, current_index + len(op.qubits))
-            all_qubits.extend(op.qubits)
-            current_index += len(op.qubits)
-        indexed_sample = self.sample(all_qubits, repetitions, seed=seed)
-
-        results = {}
-        for k, (s, e) in bounds.items():
-            before_invert_mask = indexed_sample[:, s:e]
-            results[k] = before_invert_mask ^ (np.logical_and(
-                before_invert_mask < 2, meas_ops[k].full_invert_mask()))
+            if key in seen_measurement_keys:
+                raise ValueError(f'Duplicate MeasurementGate with key {key}')
+            seen_measurement_keys.add(key)
+
+        # Find measured qubits, ensuring a consistent ordering.
+        measured_qubits = []
+        seen_qubits: Set[cirq.Qid] = set()
+        for op in measurement_ops:
+            for q in op.qubits:
+                if q not in seen_qubits:
```

Comment:
This is what I initially did, and it resulted in several fixed-seed tests being flaky. The problem is it results in a non-deterministic ordering because `set` doesn't guarantee insertion order ordering.